\documentclass[../main/thesis.tex]{subfiles}

\begin{document}

\chapter{Background}
\ifdefined\main
\acresetall
MAIN IS TRUE
\newcommand{\codePath}{../2_background/code/}
\newcommand{\figPath}{../2_background/figures/}
\else
MAIN IS NOT TRUE
\input{../notmain.tex}
\fi


\section{Cell and Molecular biology}
The most basic unit for a living organism is a \textit{cell}, often called the 'building blocks of life'.
Each living thing consists of one or more cells.
Cells can have different shapes, sizes and functions, but they all have some things in common.
Each cells consists of cytoplasm surrounded by a membrane.
This membrane is the boundary between the exterior and the inside of the cell and acts as a filter to allow certain molecules to enter or exit the cell.
A cell can grow by taking nutrients from the environment, and using them to create other molecules, or letting them interact with existing molecules \cite{bioinformatics}.
Cells can also reproduce when there are enough components in the original call to produce a duplicate cell.
Many processes in the cell are driven by \textit{proteins}.
Proteins are long chains of \textit{amino acids}.
These amino acids are joined by peptide bonds to form polypeptides.
When these polypeptides are folded by the forces between the atoms, it is called a protein.
The exact form of the protein is crucial to its function \cite{protein_misfolding}.
Examples of protein functions are breaking down molecules like other proteins, fat or carbyhydrates, fighting off foreign particles like viruses or bacteria, and assisting in a chemical reaction, in which case the protein is called an \textit{enzyme} \cite{protein_function}.
Enzymes bind to the reagents of the reaction to lower the activation energy and increase the speed of the reaction, but they are not consumed during the reaction and can be reused.
Enzymes are usually highly specific, meaning that they will only catalyze certain reactions \cite{enzyme_specificity1}\cite{enzyme_specificity2}.

It is clear that proteins play a huge role in sustaining a cell.
The information needed to create proteins is stored in \textit{DNA} or deoxyribonucleic acid.
DNA can be \textit{translated} to create new proteins (this proces is called \textit{genetic expression}), or \textit{replicated} to allow for reproduction.
A section of DNA that codes a certain protein is called a \textit{gene}.
A DNA molecule consists of two long chains of nucleotides (also called strands), which are intertwined with eachother in a double helix.
Each nucleotide is composed of a nucleobase, a sugar called deoxyribose and a phosphate group \cite{dna_structure}.
The four different nucleobases are: adenine (A), thymine (T), cytosine (C) and guanine (G).
The bases of one strand bond with bases in the opposite strand, but adenine can only bond with thymine, and cytosine only with guanine.

\figC{width=\textwidth}{dna_structure.jpeg}{Structure of DNA \cite{dna_structure_fig}}{fig:dna_structure}

Each strand has two ends: the \textit{5'-end} (pronounced five-prime) and the \textit{3'-end}.
The 5'-positions can bind a phosphate group, the 3'-positions can bind a sugar.
This leads to a \textit{backbone} of alternating sugar and phosphate groups.
The two strands are \textit{antiparallel}, this means one strands 5'-end is matched to the others 3'-end.
The orientation of the strand is significant: replication and translation can only be done in 5'$\rightarrow$3' direction.

Creating new proteins from DNA is done via messenger RNA or \textit{mRNA}, which looks like DNA, but has only one strand.
It has the same structure as DNA, but thymine (T) is replaced by uracil (U), and the sugar is \textit{ribose} instead of deoxyribose.
When reading the DNA, three bases are considered together, and are called a \textit{codon}.
Each codon describes an amino acid, but since there are only 20 different amino acids, and $4^3=64$ different codons, multiple codons map to the same amino acid, and some codons have special functions, like start (ATG) and stop (TAA, TAG, TGA) \cite{codons}.

The process starts with the enzyme \textit{RNA polymerase} binding to the correct place on the DNA with the help of a \textit{promotor}, which indicates the start of a gene.
The mRNA strand will look like the \textit{coding (or sense) strand} of the DNA, the other DNA strand is called the \textit{template (or antisense) strand}.
The RNA polymerase create a so-called \textit{transcription bubble}, which is a short section of seperated DNA, where the RNA polymerase can access the bases of the coding strand.
The polymerase now adds RNA nucleotides to the template strand, creating an mRNA strand.
This strand is seperated from the DNA template strand, allowing the DNA strands to join back together.
The mRNA strand now looks like the DNA coding strand, with the exception that thymine is replaced by uracyl.

The mRNA strand is now ready to be translated into a protein, this is done by a \textit{ribosome}, a very complex molecule consisting of several ribosomal RNA molecules and dozens of proteins \cite{DNA_translation1}.
The ribosome binds to the start of the mRNA strand.
The codons of the mRNA strand are interpreted by \textit{transfer RNA} (tRNA).
The first tRNA molecule binds to the start codon (AUG), and always carries the amino acid methionine.
The ribosome then keeps finding tRNA molecules that fit the mRNA codons, and adds their amino acids to the growing chain, creating the protein.
The tRNA molecules are not consumed during the process and can, after picking up a new amino acid, be reused.
There are no tRNA molecules that can fit a stop codon (UAA, UAG or UGA).
So instead of a tRNA molecule, a protein from a group called 'release factors' binds to the mRNA, and the ribosome releases the protein \cite{DNA_translation2}.
The used mRNA molecules degrade after they are used \cite{mRNA_degradation}.

The protein is now a chain of amino acids, linked together by peptide bonds.
This is called its \textit{primary structure} \cite{protein_structures}.
Most amino acids are nonpolar, this means their electrical charge is zero and balanced.
Others have positive or negative charges, or have no charge, but do have a dipole, these are called polar.
Polar amino acids can form hydrogen bonds \cite{hydrogen_bond}, charged amino acids can form ionic bonds \cite{ionic_bond}.
Hydrophobic amino acids can form weaker van der Waals bonds.
Most of these bonds noncovalent, which means they do not share electrons \cite{ionic_bond}.
Cysteine is the only amino acid that can form covalent bonds \cite{protein_structures}.

Bonds between parts of the protein can cause folding patterns to appear.
The most occurring types are \textit{alpha helices} and \textit{beta sheets}, shown in Figure \ref{fig:alpha_helix_beta_sheet}.
These patterns form the \textit{secondary structure} of a protein.
When these patterns interact with eachother, for example due to van der Waals bonds, the \textit{tertiary structure} forms.
Finally, a protein consisting of multiple chains, or subunits, is called the \textit{quaternary structure} \cite{protein_structures}.

\figC{scale=.3}{alpha_helix_beta_sheet.jpg}{Alpha helix and beta sheet visualized, source: \cite{alpha_helix_beta_sheet}}{fig:alpha_helix_beta_sheet}

A partially folded protein can interact with different molecules in the cell, causing improper folding.
Misfolded proteins can also clot together with other molecules, causing large aggregates.
To prevent this, \textit{chaperone proteins} surround a protein during the folding process.
Many chaperone proteins are \textit{heat shock} proteins, because heat makes proteins less stable.
The cells produces more heat shock proteins when it is exposed to heat \cite{chaperone_protein}.


\section{DNA sequencing}



\subsection{Sanger sequencing}
Sanger sequencing, also called chain-termination sequencing, is the first major DNA sequencing technique.
It is published in 1977 \cite{Sanger}, and later improvements led to the development of commercial DNA sequencing machines in 1991 \cite{sequencing_history}.
It is also the technique used by the Human Genome Project \cite{sanger_sequencing2}, a project to sequence the whole genome of a human.
The project start around 1990, and finished in 2003 \cite{human_genome_project}.
Since Sanger sequencing has a maximum chain length of about 900 basepairs \cite{sanger_sequencing2}, the resulting pieces of DNA had to be assembled after sequencing.

To sequence a DNA sample, it is mixed in a tube with: a \textit{primer}, DNA polymerase, normal DNA nucleotides (dATP, dTTP, dGTP and dCTP), and dye-labeled, chain-terminating dideoxynucleotides.
The primer is a piece of DNA that can fit onto a specific spot of a DNA strand.
Primers usually contains 18 to 24 bases \cite{sanger_primer_size}.
The chain-terminating nucleotides are ddATP, ddTTP, ddGTP and ddCTP.
They consist of a normal nucleotide, but with an oxygen atom removed, as shown in Figure \ref{fig:ddNTP}.
This small difference means that no new nucleotide can be attached to it, thus ending the chain.
The mixture is heated to split the two DNA strands.
Once the primer is bound, the DNA polymerase can start adding nucleotides to form a chain.
At one point, a dideoxynucleotide is added, and the chain cannot grow anymore.
The distribution of chain lengths is determined by the ratio of normal and dideoxynucleotides \cite{sanger_ratio}.

After a certain period, most of the dideoxynucleotides have terminated a chain, and the next phase can begin.
The chains are sorted by length by means of \textit{cappilary electrophoresis} (CE).
During CE, the chains are run through a long glass capillary filled with a gel polymer.
An electrical field is applied and the DNA fragments move through the capillary.
The speed of the fragment is inversely proportional to its weight, so the shortest chains arrive earliest at the end.
The resolution is high enough to seperate chains that differ in length by one base.
A laser excites the dye-labeled dideoxynucleotides, which emits a light at a characteristic wavelength.
These lights are detected and interpreted to get the actual DNA sequence.

\figC{scale=.5}{ddNTP.png}{Normal nucleotides and dideoxynucleotides, source: \cite{sanger_sequencing2}}{fig:ddNTP}
\figC{scale=.5}{sanger_sequencing.png}{Workflow of Sanger sequencing, source: \cite{sanger_sequencing2}}{fig:sanger_sequencing}

\subsection{Next Generation Sequencing}
Next Generation Sequencing (NGS) involves a number of different techniques.
They are different from the previous Sanger sequencing in that they are massively parallel, have high throughput at a much lower cost \cite{ngs_history}.

% Pyro
% SOLiD
% Illumina, reversible dye terminators

The first NGS method is \textit{pyrosequencing}, developed in 2005 \cite{history_sequencing}.
Like Sanger sequencing, it is a \textit{sequence-by-synthesis} method, because it relies on DNA polymerase to recreate the DNA.
The DNA strand is split, and fragmented to pieces, which are attached to microscopic beads.
The strands are cloned using Polymerase Chain Reaction (PCR) \cite{dna_cloning}\cite{ngs_sequencing_atdbio}, such that each bead has about 10 million identical copies of its fragment \cite{pyro_sequencing2}.
Each bead is place into a separate well, and each well is given a mixture that contains DNA polymerase, adenosine phosphosulfate (APS), ATP sulfurylase, luciferin, and luciferase.
ATP sulfurylase is an enzyme that combines APS and pyrophosphate (PPi) into ATP, an energy carrier.
For one cycle, one type of nucleotide is added to each of the wells.
When a nucleotide is added to the chain by DNA polymerase, it releases PPi.
This is converted to ATP by ATP sulfurylase, this energy is used by luciferase to oxidize luciferin and create light.
The amount of light is proportional to the amount of nucleotides added, and since it is known which type of nucleotide is added, the bases can be read.
The wells are cleaned by the enzyme apyrase, which degrades ATP and the unused nucleotides.
Now, the next type of nucleotide can be added.
The whole sequencing process is shown in Figure \ref{fig:pyro_sequencing}.

The newest pyrosequencing machines can produce readlenghts up to 700 basepairs.

\figC{scale=.5}{pyro_sequencing.jpg}{Workflow of pyrosequencing, source: \cite{pyro_sequencing_fig}}{fig:pyro_sequencing}

The most popular NGS method is created by Illumina \cite{most_popular_ngs}, released in 2007.
It is similar to Sanger sequencing, but uses \textit{reversible terminators}.
Another major difference is that instead of emulsion PCR, bridge PCR is used, which is more efficient \cite{ngs_sequencing_atdbio}.
Bridge PCR is further explained in Figures \ref{fig:bridge_pcr} and \ref{fig:bridge_pcr_comp}.
At the end of the PCR step, the reverse strands are washed away \cite{bridge_pcr_reverse}.
The reversible terminators are regular nucleotides, but fluorescent dye occupies the 3' end.
First, the reversible terminators are added, and bound by the DNA polymerase.
The unused nucleotides are washed away, and the dye of the bound nucleotides is released by an enzyme, allowing the bases to be read, and the chain to be extended during the next cycle.

Reads produced by this method are shorter (about 100 basepairs \cite{ngs_history}\cite{pyro_sequencing1}).

\figC{scale=.5}{bridge_pcr.png}{Process of bridge PCR, source: \cite{ngs_sequencing_atdbio}}{fig:bridge_pcr}
\figC{scale=.5}{bridge_pcr_comp.jpg}{(a) emulsion PCR, (b) bridge PCR, (c) emulsion PCR results in beads with cloned DNA strands, each well can fit one bead, (d) bridge PCR results in clusters of cloned DNA strands, source: \cite{sequencing_history}}{fig:bridge_pcr_comp}



The third major NGS method is Sequencing by Oligonucleotide Ligation and Detection (SOLiD), released by Applied Biosystems Instruments (ABI), which later became Life Technologies \cite{history_sequencing}.
SOLiD relies on ligation to chain nucleotides.
DNA ligase is able to ligate double-stranded DNA \cite{ngs_sequencing_atdbio}, whereas DNA polymerase only adds one base to a growing chain, complementary to a template strand \cite{dna_replication}.

A single strand DNA piece, known as an adapter, is merged with the template strand on one side, and attached to a bead on the other.
A primer of length N is connected to the adapter.
Now the DNA ligase can add oligonucleotides, these are strands of eight nucleotides, with fluorescent dye at the 5' end, shown in Figure \ref{fig:solid_color_coding}.
The first two bases have to be complementary to the bases in the template strand.
The degenerate and universal bases are not used.
Silver ions are used to cleave the link between bases five and six, allowing the dye to be observed.
The cleave leave a normal 5' end, which can be used to bind new oligonucleotides.
After the first round of sequencing, the new chain is removed, and a second round is started, but with a primer of length N-1.
Multiple rounds with primers with different lengths ensure that each bases is measured twice.
After the rounds, the colors can be decoded to get the actual bases, like in Figure \ref{solid_color_decoding}.
The total workflow in shown in Figure \ref{fig:solid_workflow}.

Because each base is measured twice, the accuracy can go up to 99.999\% with six primers, which is the highest of the NGS methods \cite{ngs_sequencing_atdbio}.
A downside is that the read lengths are usually short, 50 to 75 basepairs \cite{ngs_history}.

\figC{scale=.5}{solid_color_coding.jpg}{Color coding of SOLiD sequencing results, source: \cite{solid_color_coding}}{fig:solid_color_coding}
\figC{scale=.3}{solid_color_decoding.png}{Color decoding of SOLiD sequencing results, source: \cite{solid_color_decoding}}{fig:solid_color_decoding}
\figC{scale=.5}{solid_sequencing.png}{Workflow of SOLiD sequencing, source: \cite{ngs_sequencing_atdbio}}{fig:solid_workflow}

There are two other popular NGS techniques: DNA Nanoball Sequencing (DNBS) and Ion Torrent.
DNBS relies on Rolling Circle Amplification to clone the DNA strands that need to be sequenced.
The cloned strands are sequenced like in SOLiD sequencing \cite{ngs_history}.
Ion Torrent is very similar to pyrosequencing, but uses hydrogen ions instead of fluorescent dye to detect bases \cite{ngs_sequencing_atdbio}.
The number of ions is proportional to the number bases that was attached, however, at large repeating sections the accuracy decreases.
Ion Torrent produces reads of up to 400 basepaires, but the error rate is relatively high at 1.5\%.


\subsection{Long read sequencing}




\section{GPU processing}

%TODO also talk about OpenCL?
%TODO why GPU in the first place? What do they look like, what are they good at?
%TODO CUDA was released in 2006, source: programming guide

A GPU is a Graphics Processing Unit, it is a processor that is mainly used to perform video processing.
This type of processing often includes rotation and translation of objects in a space, calculating shadows and rendering images to display on a monitor.
They contain many cores that allow it to perform parallelizable tasks very quickly.
A GPGPU, or General Purpose GPU can be programmed to perform tasks that are different from video processing.
An algorithm like matrix addition is easily parallelized by assigning a matrix cell to each thread.
Each thread can perform the addition in parallel, let this take one cycle.
A sequential implementation would have taken $N\times M$ cycles, where $N$ and $M$ are the dimensions of the matrices.

CPUs usually have large caches and a complex instruction set and execution that includes out-of-order execution and branch prediction \cite.

GPUs cannot operate on their own, they must be guided by a CPU.
A general workflow using a GPU is shown in Figure \ref{gpu:workflow}.

%TODO include gpu:workflow

\begin{itemize}
\item Copy data from CPU to GPU
\item Let GPU process the data
\item Copy results from GPU to CPU
\end{itemize}

The functions that run on a GPU are called \textit{kernels}, and are usually launched by a CPU.
Kernels can also be called from other kernels.


%TODO maybe divide into subsections
\section{CUDA}
CUDA is a parallel computing platform that allows people to use Nvidia GPUs for their own applications.
Developers can write functions that will execute on the GPU called \textit{kernels}, these are launched from a CPU function.
The GPU is referred to as \textit{device} and the CPU as \textit{host}.
Kernels for CUDA are written in C++.

Kernels can be launched from the CPU in a \textit{grid} with a certain number of thread blocks or \textit{blocks} and \textit{threads}.
A grid is one-/two- or three-dimensional, and has an array of blocks in each direction.
Each block itself is also one-/two- or three-dimensional, and has an array of threads in each direction.
Figure \ref{cuda:hierarchy} shows a full grid.

%TODO picture of hierarchy \label{cuda:hierarchy}
% possible source: http://international.download.nvidia.com/pdf/kepler/NVIDIA-Kepler-GK110-GK210-Architecture-Whitepaper.pdf

Each thread executes the kernel code, although they usually operate on different data.
Threads in a block can communicate via shared memory.

On a hardware level, an NVIDIA GPU is divided into Streaming Multiprocessors (SMX).
Each SMX contains a number of cores, or Streaming Processors (SP), these are the basic building blocks and perform the actual calculations.
Each block is assigned to at most one SMX.
This block's threads are then executed as warps, with 32 threads per warp.
Each SMX has multiple warp schedulers, so multiple warps can run in parallel on an SMX.
All threads in a warp must execute the same instruction, if a thread is the only to take a branch, the other threads must wait until the branch is completed, this is called divergence.
Memory operations are also executed in parallel, this means that all threads try to read/write to the memory in parallel.
If the addresses are next to eachother, only one memory transaction is needed, since a transaction processes a whole memory line.
This is known as coalescing.

%TODO include figure of coalescing


%TODO include figure of hardware
%https://www.techpowerup.com/img/12-05-17/155a.jpg

%Tesla-Fermi-Kepler-Maxwell-Pascal-Volta


\subsection{Memory hierarchy}
GPUs have several different memory types and levels.
Not all of these are accessible from the host or from other components of the memory hierarchy \cite{cuda}. Figure \ref{fig:memory_hierarchy} shows three types of memory.
\begin{itemize}
\item Global memory: The largest and slowest memory, located outside the chip. Can be read and written from the host. Best used for coalesced operations.
\item Local memory: Each thread has private memory for when registers are not enough. It is stored in global memory, so accesses are very slow.
\item Registers: The fastest memory available, located on-chip. A thread has a maximum number of registers available depending on Compute Capability.
\item Shared memory: Each SMX has shared memory, it can be accesses by every thread in every block on the SMX. This can be used by threads in a block to communicate. It is located on-chip, so very fast.
\item Constant memory: The host can initialize this memory, the kernel can only read it. Reading is as fast as reading a register, but only when all addresses of a half-warp are the same, otherwise reading is serialized. It resides off-chip, but is cached on-chip.
\item Texture memory: Can only be written from the host. Resides off-chip, but is cached on-chip, like constant memory. The main feature about the texture memory is that it is cached for 2D spatial locality. Figure \ref{fig:texture_cache} shows an access pattern that would not be cached with a typical scheme.
\item L2 cache: Behaves as cache for device memory and is shared among all SMXs. The cache line size is 32B.
\item L1 cache: Cache line size is 128B. Its default behaviour is to only cache local memory, not global memory. However, for certain architectures applications can opt-in to cache both global and local loads in L1 \cite{L1cache}.
\end{itemize}

\figC{scale=.5}{memory_hierarchy.png}{Memory hierarchy \cite{cuda}}{fig:memory_hierarchy}

%\includegraphics[scale=.2]{\fig texture_cache.png}
\figC{scale=.2}{texture_cache.png}{This access pattern has spatial locality, and could be cached in the texture cache \cite{texture_cache}}{fig:memory_hierarchy}

\subsection{Streams}
When copying data to or from the GPU, there are usually no kernels running in a naive implementation.
This means that if the copy time is large, the overall efficiency is quite low.
These operations can be overlapped or pipelined by using streams. Figure \ref{fig:streams} shows how the execution time can decrease by overlapping copying and computing.

\figC{scale=.2}{streams.png}{Different versions of the same computation \cite{streams}}{fig:streams}





%TODO include this section? this work mainly focuses on GPU computation, instead of heterogeneous computation
\section{Heterogeneous computing}
%\cite{hetero_survey}




\end{document}










